### PR TITLE
Add support for Python SQL connections and Python DuckDB implementation

### PR DESCRIPTION
**Pull request by jonmmease**
## Overview
This PR introduces the ability to define SQL connections in Python. To prove out the idea, a Python DuckDB connection is implemented.

### Python SqlConnection
An `SqlConnection` Python abstract class is defined in `python/vegafusion/vegafusion/connection/__init__.py`. Subclasses are required to implement three methods:
 - `tables(self) -> Dict[str, pa.Schema]` to return a dict from registered table names to pyarrow schemas.
 - `dialect(cls) -> str` to return the name of the SQL dialect that the connection expects. These correspond to the dialects defined in the `vegafusion-sql` crate.
 - `fetch_query(self, query: str, schema: pa.Schema) -> pa.Table` to evaluate the requested SQL query and return the result as a pyarrow Table.  The requested query may reference named tables returned by the `tables` method. The expected result schema is provided.

SQL connections to read-only databases would typically implement only these three methods (Though we don't have an example of this yet).  Additional optional methods may be overridden by connections (like duckdb) that support dynamic registration of tables from in-memory data structures.

For example:
 - `register_arrow(self, name: str, table: pa.Table, temporary: bool = False)` may be implemented to support the dynamic registration of a pyarrow Table for use in `fetch_query`. 

In `vegafusion-python-embed/src/connection.rs`, a Rust `PySqlConnection` struct is defined that implements the Rust `Connection` and `SqlConnection` traits. `PySqlConnection` wraps a Python object (through PyO3) that is a subclass of the Python `SqlConnection` abstract class.

### Fallback
Python SQL connections may optionally declare that they should fall back to the built-in DataFusion connection when SQL is encountered that is not supported by the connection's dialect. Fallback defaults to True.

### Multi-threading
To work around deadlocking the Python interpreter, it was necessary to configure the tokio runtime in single-threaded mode when a Python SqlConnection is registered with the VegaFusion runtime.  We still use a multi-threaded tokio runtime for the pure Rust DataFusion connection.

## Usage
Here is an example of using the DuckDB connection from Python

```python
import vegafusion as vf
import pandas as pd
vf.runtime.set_connection("duckdb")
vf.enable()
```
```python
import pandas as pd
import altair as alt

flights = pd.read_parquet(
    "https://vegafusion-datasets.s3.amazonaws.com/vega/flights_1m.parquet"
)

delay_hist = alt.Chart(flights).mark_bar().encode(
    alt.X("delay", bin=alt.Bin(maxbins=30)),
    alt.Y("count()")
)
delay_hist
```
visualization

## Performance
Most significant benefit of the DuckDB connection is its ability to bind tables directly to Pandas DataFrames without serialization to arrow.  This results in significant speedups with the mime renderer and the `vf.transformed_data` function. I have not seen much benefit in the widget renderer, since in this case the DataFrame is always serialized to arrow and written to disk.  In this case, my anecdotal observation is that the DuckDB and DataFusion connections have very similar performance.

Here's an example of the speedup when calling `transformed_data` on a histogram chart of a 1 million row movies dataset.

```python
import vegafusion as vf
import pandas as pd
import altair as alt

# Load large movies dataset
source = "https://vegafusion-datasets.s3.amazonaws.com/vega/movies_1m.parquet"
movies = pd.read_parquet(source)

chart = alt.Chart(movies).mark_bar().encode(
    alt.X("IMDB_Rating", bin=True),
    alt.Y("count()")
)
``` 
First with the DataFusion connection
```
vf.runtime.set_connection("datafusion")
%time vf.transformed_data(chart)
```
```
CPU times: user 656 ms, sys: 136 ms, total: 792 ms
Wall time: 731 ms
```

Now with the DuckDB connection
```python
vf.runtime.set_connection("duckdb")
%time vf.transformed_data(chart)
```
```
CPU times: user 265 ms, sys: 8.81 ms, total: 274 ms
Wall time: 62.5 ms
```
Over 10x faster! 🎉 

## Limitations
The DuckDB connection does have some limitations

 - The DuckDB CSV reader sometimes has trouble loading CSV files from the jsdelivr URLs in the vega datases package. Sometimes it results in an error complaining that the response does not include Content Length. It's working on the moment so I don't have the exact error at hand. This PR changes some of these URLs to equivalent URLs that pull files from GitHub where there is no issue.
 - DuckDB doesn't have a function to automatically parse a wide variety of timestamp formats. Vega relies on the browser's `Date` class to parse a variety of date formats, and DataFusion defines a custom UDF to match this behavior. Because of this, the DuckDB connection will fail with date parsing errors when non ISO-8601 timestamp strings are used as timestamps.
 - Relatedly, DataFusion follows an idiosyncrasy of the browser's timestamp parsing logic. timestamp strings are always parsed in the local timezone unless they are of the form YYYY-MM-DD, in which case they are parsed in UTC. The DuckDB connection will consistently parse all supported timestamp strings in the local timezone. 

None of these limitations are relevant when using a Pandas DataFrame as the data source for a chart (as long as timestamp strings are already converted to timestamps), but they come up when loading data from external URLs, as is the case in a lot of the Altair gallery examples.  This is why I don't foresee changing the default connection to DuckDB in the near future.

### CODE REVIEW

File: vegafusion-dataframe/src/dataframe.rs [hunk 1]
```diff
@@ -36,15 +39,15 @@ pub trait DataFrame: Send + Sync + 'static {
             .with_context(|| String::from("Failed to concatenate RecordBatches"))
     }
 
-    fn sort(&self, _expr: Vec<Expr>, _limit: Option<i32>) -> Result<Arc<dyn DataFrame>> {
+    async fn sort(&self, _expr: Vec<Expr>, _limit: Option<i32>) -> Result<Arc<dyn DataFrame>> {
```
Comment (jonmmease) [inline vegafusion-dataframe/src/dataframe.rs:42]: query methods had to be made async to support fallback calculation

File: vegafusion-runtime/src/data/tasks.rs [hunk 2]
```diff
@@ -245,14 +251,14 @@ lazy_static! {
     .collect();
 }
 
-const DATASET_CDN_BASE: &str = "https://cdn.jsdelivr.net/npm/vega-datasets";
-const DATASET_TAG: &str = "v2.2.0";
+const DATASET_BASE: &str = "https://raw.githubusercontent.com/vega/vega-datasets";
```
Comment (jonmmease) [inline vegafusion-runtime/src/data/tasks.rs:254]: duckdb  has trouble loading files from jsdelivr, more reliable to pull from GitHub

File: vegafusion-runtime/tests/specs/custom/stocks_pivot.comm_plan.json
Comment (jonmmease) [inline vegafusion-runtime/tests/specs/custom/stocks_pivot.comm_plan.json:1]: Test file was not being used

File: vegafusion-sql/Cargo.toml [hunk 3]
```diff
@@ -17,6 +17,10 @@ rstest_reuse = "0.5.0"
 lazy_static = "^1.4.0"
 toml = "0.7.2"
 
+[dev-dependencies.async-std]
+version = "1.12.0"
+features = ["attributes"]
```
Comment (jonmmease) [inline vegafusion-sql/Cargo.toml:22]: This is to support rstest on async test functions

File: vegafusion-sql/src/compile/expr.rs [hunk 4]
```diff
@@ -367,7 +367,15 @@ impl ToSqlExpr for Expr {
                         let units = match window_frame.units {
                             WindowFrameUnits::Rows => SqlWindowFrameUnits::Rows,
                             WindowFrameUnits::Range => SqlWindowFrameUnits::Range,
-                            WindowFrameUnits::Groups => SqlWindowFrameUnits::Groups,
+                            WindowFrameUnits::Groups => {
+                                if dialect.supports_window_frame_groups {
+                                    SqlWindowFrameUnits::Groups
+                                } else {
+                                    return Err(VegaFusionError::sql_not_supported(
```
Comment (jonmmease) [inline vegafusion-sql/src/compile/expr.rs:374]: Turns out only Postgres and DataFusion support window frame GROUPS

File: vegafusion-sql/src/dialect/transforms/str_to_utc_timestamp.rs [hunk 5]
```diff
@@ -32,7 +32,7 @@ fn process_str_to_utc_timestamp_args(
     Ok((sql_arg0, time_zone))
 }
 
-/// Convert str_to_utc_timezone(ts, tz) ->
+/// Convert str_to_utc_timestamp(ts, tz) ->
```
Comment (jonmmease) [inline vegafusion-sql/src/dialect/transforms/str_to_utc_timestamp.rs:35]: Just fixing the function name

File: vegafusion-sql/tests/test_aggregate.rs [hunk 6]
```diff
@@ -138,57 +143,59 @@ mod test_variance_aggs {
...
+                            filter: None,
+                        })
+                        .mul(lit(100)),
+                    )
+                    .div(lit(100))
+                    .alias("var_pop_a"),
+                ],
+            )
+            .await;
+        let df_result = if let Ok(df) = df_result {
```
Comment (jonmmease) [inline vegafusion-sql/tests/test_aggregate.rs:198]: The `and_then` pattern we used before doesn't play well with async query methods, so restructure to use an `if let Ok` pattern.

**Comment by sb-eip** [comment]
Impressive speedup! Looking forward to trying it out :) Thanks @jonmmease!

**Comment by jonmmease** [comment]
Thanks @binste!  I'll definitely be interested to hear how it works for you.  This functionality was released in version 1.1.0rc1 this morning, so feel free to give it a spin!

For your use case, it sounds like the workflow added in https://github.com/hex-inc/vegafusion/pull/252 might be a good fit, where you reuse an existing DuckDB connection and refer to tables using `table://...` URLs.

**Pull request merged**